src/linux/write_zeroes.rs

```diff
@@ -32,150 +35,276 @@ impl PunchHole for File {
+    ///
+    /// This method will continuously call `write_zeroes` until the requested `length` is satisfied
+    /// or an error is encountered.
+    fn write_zeroes_all(&mut self, mut length: usize) -> Result<()> {
+        while length > 0 {
+            match self.write_zeroes(length) {
+                Ok(0) => return Err(Error::from(ErrorKind::WriteZero)),
+                Ok(bytes_written) => {
+                    length = length
+                        .checked_sub(bytes_written)
+                        .ok_or_else(|| Error::from(ErrorKind::Other))?
+                }
+                // If the operation was interrupted, we should retry it.
+                Err(e) => {
+                    if e.kind() != ErrorKind::Interrupted {
+                        return Err(e);
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
 }
 
-impl<T: PunchHole + Seek + Write> WriteZeroes for T {
-    fn write_zeroes(&mut self, length: usize) -> Result<usize> {
-        // Try to punch a hole first.
-        let offset = self.seek(SeekFrom::Current(0))?;
-        if let Ok(()) = self.punch_hole(offset, length as u64) {
-            // Advance the seek cursor as if we had done a real write().
-            self.seek(SeekFrom::Current(length as i64))?;
-            return Ok(length);
+/// A trait for writing zeroes to an arbitrary position in a file.
+pub trait WriteZeroesAt {
+    /// Write up to `length` bytes of zeroes starting at `offset`, returning how many bytes were
+    /// written.
+    fn write_zeroes_at(&mut self, offset: u64, length: usize) -> Result<usize>;
+
+    /// Write zeroes starting at `offset` until `length` bytes have been written.
+    ///
+    /// This method will continuously call `write_zeroes_at` until the requested `length` is
+    /// satisfied or an error is encountered.
+    fn write_zeroes_all_at(&mut self, mut offset: u64, mut length: usize) -> Result<()> {
+        while length > 0 {
+            match self.write_zeroes_at(offset, length) {
+                Ok(0) => return Err(Error::from(ErrorKind::WriteZero)),
+                Ok(bytes_written) => {
+                    length = length
+                        .checked_sub(bytes_written)
+                        .ok_or_else(|| Error::from(ErrorKind::Other))?;
+                    offset = offset
+                        .checked_add(bytes_written as u64)
+                        .ok_or_else(|| Error::from(ErrorKind::Other))?;
+                }
+                Err(e) => {
+                    // If the operation was interrupted, we should retry it.
+                    if e.kind() != ErrorKind::Interrupted {
+                        return Err(e);
+                    }
+                }
+            }
         }
+        Ok(())
+    }
+}
 
-        // fall back to write()
+impl WriteZeroesAt for File {
+    fn write_zeroes_at(&mut self, offset: u64, length: usize) -> Result<usize> {
+        // Try to use fallocate() first.
```

Comment:
Let's also say that we are trying to use fallocate first because it is more efficient that writing zeroes.

Reply:
good idea